### PR TITLE
Draft: Add UPDR algorithm

### PR DESCRIPTION
**Pull request by oralmer**
Added an initial UPDR implementation based on the one in mypyvy. Created as a project in the course "Automatic verification of systems" taught by Prof. Sharon Shoham Buchbinder. 

The implementation itself takes 11.31 secs to run on `lockserver.fly` as opposed to the 5.47 secs it takes mypyvy to run updr on `lockserv.pyv`. Sadly, when attempting more complex problems, such as `consensus.fly` , we get a timeout. This implementation will require some more work until it can be merged IMO.

The PR also contains some helper functions that the mypyvy implentation used and were added to flyvy:
 - Conversion to CNF.
 - Splitting a CNF term into clauses.
 - Negation that pushes the `-` inward in a term.
 - Removal of `prime` from a term.

If any of these functions feel useful enough, I can open a separate PR for them.

### CODE REVIEW

**Comment by wilcoxjay** [comment]
I cleaned up the the git history and rebased this on top of main. We should be good to merge it for now, and we can continue to iterate on the code.

File: tests/test_examples.rs [hunk 1]
```diff
@@ -21,7 +21,7 @@ use serde_derive::Deserialize;
 use temporal_verifier::solver::solver_path;
 use walkdir::WalkDir;
 
-const SOLVERS_TO_TEST: [&str; 3] = ["z3", "cvc4", "cvc5"];
+const SOLVERS_TO_TEST: [&str; 1] = ["z3"];
```
Comment (tchajed) [inline tests/test_examples.rs:24]: This needs to be reverted.

**Review by tchajed — CHANGES_REQUESTED** [review]
`test_examples.rs` should still test all solvers.

We're going to merge this and work on cleaning up `inference/basics.rs` (and `updr.rs`) later, so I haven't gone through and added comments to those files.

File: tests/test_examples.rs (hunk 1 above)
Comment (wilcoxjay) [reply tests/test_examples.rs]: good catch! fixed

**Pull request merged**